api/v1alpha1/environment_webhook.go

```diff
@@ -0,0 +1,95 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var environmentlog = logf.Log.WithName("environment-resource")
+
+func (r *Environment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-environment,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=environments,verbs=create;update,versions=v1alpha1,name=menvironment.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &Environment{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *Environment) Default() {
+
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-environment,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=environments,verbs=create;update,versions=v1alpha1,name=venvironment.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &Environment{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Environment) ValidateCreate() error {
+	environmentlog = environmentlog.WithValues("controllerKind", "Environment").WithValues("name", r.Name).WithValues("namespace", r.Namespace)
+	environmentlog.Info("validating the create request")
+
+	return r.validateIngressDomain()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *Environment) ValidateUpdate(old runtime.Object) error {
+	environmentlog = environmentlog.WithValues("controllerKind", "Environment").WithValues("name", r.Name).WithValues("namespace", r.Namespace)
+	environmentlog.Info("validating the update request")
+
+	return r.validateIngressDomain()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *Environment) ValidateDelete() error {
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+// validateIngressDomain validates the ingress domain
+func (r *Environment) validateIngressDomain() error {
+	unstableConfig := r.Spec.UnstableConfigurationFields
+	if unstableConfig != nil {
+		// if cluster type is Kubernetes, then Ingress Domain should be set
+		if unstableConfig.ClusterType == ConfigurationClusterType_Kubernetes && unstableConfig.IngressDomain == "" {
+			return fmt.Errorf(MissingIngressDomain)
+		}
+
+		// We use the DNS-1123 format for ingress domain, so ensure it conforms to that specification
+		if len(validation.IsDNS1123Subdomain(unstableConfig.IngressDomain)) != 0 {
+			return fmt.Errorf(InvalidDNS1123Subdomain, unstableConfig.IngressDomain)
+		}
```

Comment:
This will return an error in the OpenShift case, won't it?

It may be worth adding a unit test to environment_webhook_unit_test.go for the openshift case, to make sure it passes if ingress domain is not set.

Reply:
oops, i missed that

Reply:
Updated. We are now validating Ingress Domain whenever it is provided, irrespective of Kubernetes or OpenShift. Since we can also create both Routes and Ingresses on OpenShift with domain and host.